sky/skylet/providers/kubernetes/node_provider.py

```diff
@@ -257,9 +257,14 @@ def create_node(self, node_config, tags, count):
                     self.namespace, service_spec)
                 new_svcs.append(svc)
 
-        # Wait for all pods to be ready, and if it exceeds the timeout, raise an
-        # exception. If pod's container is ContainerCreating, then we can assume
-        # that resources have been allocated and we can exit.
+        # Wait for all pods including jump pod to be ready, and if it
+        # exceeds the timeout, raise an exception. If pod's container
+        # is ContainerCreating, then we can assume that resources have been
+        # allocated and we can exit.
+        ssh_jump_pod_name = conf['metadata']['labels']['skypilot-ssh-jump']
+        jump_pod = kubernetes.core_api().read_namespaced_pod(
+            ssh_jump_pod_name, self.namespace)
+        new_nodes.append(jump_pod)
```

Comment:
I tried to test this by editing `kubernetes-ssh-jump.yml.j2` to request a large amount of CPU for the SSH jump pod (so that it fails to get scheduled):
```
# containers:
  ...
     resources:
       requests:
         cpu: 20
```

When I ran `sky launch -c test -- echo hi`, I got this error (though failover continued as expected, but the error message was confusing):
```
D 10-26 11:32:53 cloud_vm_ray_backend.py:2006] `ray up` takes 11.3 seconds with 1 retries.
W 10-26 11:32:53 cloud_vm_ray_backend.py:997] Got error(s) in kubernetes:
W 10-26 11:32:53 cloud_vm_ray_backend.py:999] 	sky.skylet.providers.kubernetes.config.KubernetesError: An error occurred while trying to fetch the reason for pod scheduling failure. Error: UnboundLocalError: local variable 'event_message' referenced before assignment
...
```

On the second run, I got:
```
W 10-26 11:36:14 cloud_vm_ray_backend.py:999] 	sky.skylet.providers.kubernetes.config.KubernetesError: Timed out while waiting for nodes to start. Cluster may be out of resources or may be too slow to autoscale. Pod status: PendingDetails: 'skip schedule deleting pod: default/test-2ea4-ray-head'
```

Both of these messages were while `kubectl describe pod` was giving a clear failed scheduling message:
```
Events:
  Type     Reason            Age   From               Message
  ----     ------            ----  ----               -------
  Warning  FailedScheduling  8s    default-scheduler  0/1 nodes are available: 1 Insufficient cpu. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod.
```

Compare this to our regular message when a task requests too many CPUs (e.g., sky launch --cpus 8), which clearly states that the cluster is out of CPU and suggests debugging steps. Can we have the same message for jump pod too?
```
W 10-26 11:34:57 cloud_vm_ray_backend.py:997] Got error(s) in kubernetes:
W 10-26 11:34:57 cloud_vm_ray_backend.py:999] 	sky.skylet.providers.kubernetes.config.KubernetesError: Insufficient CPU capacity on the cluster. Other SkyPilot tasks or pods may be using resources. Check resource usage by running `kubectl describe nodes`
```

Reply:
Had this error due to two reasons:
1. Was not updating the pod status by re-reading it in `_raise_pod_scheduling_errors`. 
2. Was attempting to read an event of a pod that is already scheduled and is not in `Pending` status.

Fixed the two above and setting excessive resources for jump pod or the pod instance, both fails with correct error message. Following is displayed when excessive amount of CPUs are requested for the jump pod:

```
$ sky launch --cloud kubernetes -y
I 10-30 05:57:34 optimizer.py:674] == Optimizer ==
I 10-30 05:57:34 optimizer.py:697] Estimated cost: $0.0 / hour
I 10-30 05:57:34 optimizer.py:697] 
I 10-30 05:57:34 optimizer.py:769] Considered resources (1 node):
I 10-30 05:57:34 optimizer.py:818] ---------------------------------------------------------------------------------------------
I 10-30 05:57:34 optimizer.py:818]  CLOUD        INSTANCE    vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE   COST ($)   CHOSEN   
I 10-30 05:57:34 optimizer.py:818] ---------------------------------------------------------------------------------------------
I 10-30 05:57:34 optimizer.py:818]  Kubernetes   2CPU--2GB   2       2         -              kubernetes    0.00          ✔     
I 10-30 05:57:34 optimizer.py:818] ---------------------------------------------------------------------------------------------
I 10-30 05:57:34 optimizer.py:818] 
Running task on cluster sky-xxxx-gcpuser...
I 10-30 05:57:34 cloud_vm_ray_backend.py:4382] Creating a new cluster: 'sky-xxxx-gcpuser' [1x Kubernetes(2CPU--2GB)].
I 10-30 05:57:34 cloud_vm_ray_backend.py:4382] Tip: to reuse an existing cluster, specify --cluster (-c). Run `sky status` to see existing clusters.
I 10-30 05:57:35 cloud_vm_ray_backend.py:1449] To view detailed progress: tail -n100 -f /home/gcpuser/sky_logs/sky-2023-10-30-05-57-33-829528/provision.log
I 10-30 05:57:36 cloud_vm_ray_backend.py:1884] Launching on Kubernetes 
W 10-30 05:57:48 cloud_vm_ray_backend.py:997] Got error(s) in kubernetes:
W 10-30 05:57:48 cloud_vm_ray_backend.py:999] 	sky.skylet.providers.kubernetes.config.KubernetesError: Insufficient CPU capacity on the cluster. Other SkyPilot tasks or pods may be using resources. Check resource usage by running `kubectl describe nodes`.
W 10-30 05:57:54 cloud_vm_ray_backend.py:2221] sky.exceptions.ResourcesUnavailableError: Failed to acquire resources in all zones in kubernetes. Try changing resource requirements or use another region.
W 10-30 05:57:54 cloud_vm_ray_backend.py:2230] 
W 10-30 05:57:54 cloud_vm_ray_backend.py:2230] Provision failed for 1x Kubernetes(2CPU--2GB) in kubernetes. Trying other locations (if any).
Clusters
No existing clusters.

sky.exceptions.ResourcesUnavailableError: Failed to provision all possible launchable resources. Relax the task's resource requirements: 1x Kubernetes()
To keep retrying until the cluster is up, use the `--retry-until-up` flag.
```

Reply:
Thanks for fixing this! We may also need to update `kubernetes_utils.clean_zombie_ssh_jump_pod()` to delete a pending SSH jump pod if it's out of resources. Currently, if the jump pod's CPU resources are not satisfied, then it is not cleaned up and stays indefinitely in a pending state:

```
# Right after sky launch exits, the cluster pod is terminated
(base) ➜ k get pods
NAME                    READY   STATUS        RESTARTS   AGE
sky-ssh-jump-2ea485ef   0/1     Pending       0          13s
test-2ea4-ray-head      1/1     Terminating   0          13s


# But the ssh pod remains pending forever. We should clean this up in clean_zombie_ssh_jump_pod()
(base) ➜  k get pods
NAME                    READY   STATUS    RESTARTS   AGE
sky-ssh-jump-2ea485ef   0/1     Pending   0          15s
```